src/DataStructures/Mesh.hpp

```diff
@@ -0,0 +1,149 @@
+ *
+ * \tparam Dim the number of dimensions of the computational grid.
+ */
+template <size_t Dim>
+class Mesh : public Index<Dim> {
+ public:
+  Mesh() noexcept = default;
+
+  /*!
+   * \brief Construct a computational grid with the same number of collocation
+   * points in each dimension.
+   *
+   * \param isotropic_extents The number of collocation points in each
+   * dimension.
+   * \param basis The choice of spectral basis to compute the
+   * collocation points
+   * \param quadrature The choice of quadrature to compute
+   * the collocation points
+   */
+  Mesh(const size_t isotropic_extents, const Spectral::Basis basis,
+       const Spectral::Quadrature quadrature) noexcept
+      : Index<Dim>(isotropic_extents) {
+    bases_.fill(basis);
+    quadratures_.fill(quadrature);
+  }
+
+  /*!
+   * Construct a computational grid where each dimension can have a different
+   * number of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param basis The choice of spectral basis to compute the
+   * collocation points
+   * \param quadrature The choice of quadrature to compute
+   * the collocation points
+   */
+  Mesh(std::array<size_t, Dim> extents, const Spectral::Basis basis,
+       const Spectral::Quadrature quadrature) noexcept
+      : Index<Dim>(std::move(extents)) {
+    bases_.fill(basis);
+    quadratures_.fill(quadrature);
+  }
+
+  /*!
+   * Construct a computational grid where each dimension can have both a
+   * different number and placement of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param bases The choice of spectral bases to compute the
+   * collocation points per dimension
+   * \param quadratures The choice of quadratures to compute
+   * the collocation points per dimension
+   */
+  Mesh(std::array<size_t, Dim> extents, std::array<Spectral::Basis, Dim> bases,
+       std::array<Spectral::Quadrature, Dim> quadratures) noexcept
+      : Index<Dim>(std::move(extents)),
+        bases_(std::move(bases)),
+        quadratures_(std::move(quadratures)) {}
+
+  /*!
+   * \brief The basis chosen in each dimension of the grid.
+   */
+  const std::array<Spectral::Basis, Dim>& bases() const noexcept {
+    return bases_;
+  }
+
+  /*!
+   * \brief The quadrature chosen in each dimension of the grid.
+   */
+  const std::array<Spectral::Quadrature, Dim>& quadratures() const noexcept {
+    return quadratures_;
+  }
+
+  /*!
+   * Returns a Mesh with the d-th dimension removed.
+   *
+   * \param d The dimension to remove (zero-indexed).
+   */
+  template <size_t N = Dim, Requires<(N > 0)> = nullptr>
+  Mesh<Dim - 1> slice_away(size_t d) const noexcept;
+
+  /*!
+   * \brief The total number of grid points in all dimensions.
+   *
+   * \details `DataVector`s that represent field values on the grid have this
+   * many entries.
+   */
+  size_t number_of_grid_points() const noexcept {
```

Comment:
Wouldn't this make more sense as a method on Index?

Reply:
Index has the method `product()`, since we could use it for non-grid point things. @kidder am I remembering that correctly?